### ISSUE TITLE
Implementing event listener callbacks for click, hover, etc ...

### ISSUE THREAD

**Issue by pdemilly** [hikerpig/pintora#83]
I would like to use pintora in my angular app but need to be able to catch a click event as the user selects a different element of a graph to be edited dynamically

Is that part of the plans? I have seen any code related to this in the source tree

Thanks

**Comment by hikerpig** [comment]
@pdemilly  I havn't thought about it but I think it is worth a try.
Could you exapnd this with more details ?
1. What would you expect in the event payload ?
2. What do you mean _to be edited dynamically_ ?
3. Will this require adding new grammar to the DSL ?

**Comment by pdemilly** [comment]
For example in mermaid there is the ability to define click event when using Security level loose. see [mermaid doc](https://mermaid-js.github.io/mermaid/#/flowchart?id=interaction
)

```
function clicked(id) {
  console.log(`id ${id} has been clicked`);
}
```

```mermaid
graph TD
    A-->B
    click A clicked "click A"
    click B alert "click B"
```
To answer your questions the event payload could be the node and the mouse event. What I mean by edited dynamically is to build the chart dynamically. For example clicking on a node, could pop up a drop down menu with the choice of adding a new node. It could require a change in grammar like in mermaid or could be a configuration option during initialization of the chart (probably preferable)

**Comment by hikerpig** [comment]
I see, very good advice. I should take this to the roadmap. Will take a look at it this weekend.

The event payload will include some meaningful entity, like an actor in the sequenceDiagram, or an entity in the erDiagram.

But since the event is captured in the renderer level, I need to come up with a solution to link it back to the DiagramIR level.

Not all visual elements can be easily traced back to a meaningful logic entity, say you click an arrow, it takes some work to track back to know that it belongs to an edge between two nodes.

**Comment by hikerpig** [comment]
I've opened a draft PR for this feature.

You can goto
-  [vercel preview page](https://pintorajs-git-feature-event-callback-hikerpig.vercel.app/docs/advanced/diagram-event/) to see the docs
- [live editor demo page](https://pintorajs-git-feature-event-callback-hikerpig.vercel.app/demo/live-editor/) and open the devtools console, click the 'DELIVERY' entity and see the event payload data.

Currently only er diagram will emit DiagramEvent

**Comment by hikerpig** [comment]
This is released in v0.5.2, only er and sequence diagram will emit DiagramEvent.
